File: Data Structures & Algorithms/valid-parenthesis-string/submission-0.cpp
class Solution {

int n ;

bool dfs(string& s,int idx,int open){
    if(idx == n) return open == 0 ;
    if(open < 0) return false ;
    if(s[idx] == '(') return dfs(s,idx+1,open+1) ;
    if(s[idx] == ')') return dfs(s,idx+1,open-1) ;

    return dfs(s,idx+1,open) | dfs(s,idx+1,open-1) | dfs(s,idx+1,open+1) ;
}

public:
    bool checkValidString(string s) {
        n = s.length() ;
        return dfs(s,0,0) ;
    }
};

File: Data Structures & Algorithms/count-connected-components/submission-0.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:

        def dfs(vis:List[bool],adj:List[List[int]],u:int):
            if vis[u]:
                return 
            vis[u] = True
            for v in adj[u]:
                if not vis[v]:
                    dfs(vis,adj,v)
            return 
        
        ans = 0
        vis = [False]*n
        adj = [[] for _ in range(n)]

        for u,v in edges:
            adj[u].append(v)
            adj[v].append(u)
        
        for u in range(n):
            if not vis[u]:
                ans += 1
                dfs(vis,adj,u)

        return ans
